### PR TITLE
Initial pass at enabling dom diffing for custom element templates con…

### PR DESCRIPTION
**Pull request by kristoferjoseph**
…taining slots

### CODE REVIEW

**Comment by macdonst** [comment]
@kristoferjoseph seems good to me but do we have any test cases we can add?

**Comment by kristoferjoseph** [comment]
> @kristoferjoseph seems good to me but do we have any test cases we can add?

This is what I have been working on. There are a ton of edge cases and it requires the tests to pull in all of the other mixins to test correctly.

**Comment by kristoferjoseph** [comment]
This is still very much a WIP as I keep uncovering more and more edge cases.

File: index.mjs [hunk 1]
```diff
@@ -126,9 +150,15 @@ const CustomElementMixin = (superclass) => class extends superclass {
       if (slot) {
         if (slot.name) {
           if (!namedSlots[slot.name]) namedSlots[slot.name] = { slotNode: slot, contentToSlot: [] }
+          if (child['dataset']) {
+            child.dataset.slotted = true
```
Comment (kristoferjoseph) [inline index.mjs]: Use of `dataset.slotted` to track what elements are slotted in.
Needs some additions for text content. 
These additions will also need to be added to Enhance SSR during slot expansion so this will work with SSR'd elements as well.

File: test/diff-test.html [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+        const updated = document.createElement('div')
+        updated.innerHTML = tmp
+        this.scrubTemplate(updated)
+
+        const root = this.shadowRoot
+          ? this.shadowRoot
+          : this
+
+        const slotEl = document.createElement('div')
```
Comment (kristoferjoseph) [inline test/diff-test.html]: These additions will be added to the morph mixins as well. 
Once they land in the morphdommixin I will use that in these tests.

File: test/diff-test.html [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+      render(args) {
+        return MyHeading(args)
+      }
+    }
+
+    customElements.define("my-heading", MyHeadingElement)
+
+    runTests(()=> {
+      describe('CustomElementMixin Slots', ()=> {
```
Comment (kristoferjoseph) [inline test/diff-test.html:108]: I need to add another test case for just solo text content not in a wrapping element.

**Review by ryanbethel — APPROVED** [review]
LGTM. Should we add the small fix I made in the other PR for scope=global style tags.

**Comment by kristoferjoseph** [comment]
> LGTM. Should we add the small fix I made in the other PR for scope=global style tags.

Seems like a good idea so that these changes all go out in one release

**Review by macdonst — APPROVED** [review]

File: index.mjs (hunk 1 above)
Comment (kristoferjoseph) [reply index.mjs]: I am going to try and use a `slot` attribute with no value instead of `data-slotted` so we can query for a `slot` attribute and get named slots as well.

File: index.mjs [hunk 4]
```diff
@@ -154,14 +154,26 @@ const CustomElementMixin = (superclass) => class extends superclass {
...
           }
-          namedSlots[slot.name].contentToSlot.push(child)
         } else {
           if (!unnamedSlot["slotNode"]) unnamedSlot = { slotNode: slot, contentToSlot: [] }
           if (child['dataset']) {
             child.dataset.slotted = true
+            unnamedSlot.contentToSlot.push(child)
+          }
+          else {
+            const wrapperSpan = document.createElement('span')
```
Comment (kristoferjoseph) [inline index.mjs:159]: This is not ideal but I think it will work as expected. 
I don't like that we have to add markup that the user did not author but this approach is consistent with how SSR works.
